README.md
**Status:** Expect regular updates and bug fixes.

# Tool for simulating dynamics of PV-DER
[![Build Status](https://travis-ci.org/sibyjackgrove/SolarPV-DER-simulation-utility.svg?branch=master)](https://travis-ci.org/sibyjackgrove/SolarPV-DER-simulation-utility)
![PyPI - Downloads](https://img.shields.io/pypi/dm/pvder?label=PyPI%20Downloads)
[![CodeFactor](https://www.codefactor.io/repository/github/tdcosim/solarpv-der-simulation-tool/badge)](https://www.codefactor.io/repository/github/tdcosim/solarpv-der-simulation-utility)

Solar photovoltaic distributed energy resources (PV-DER) are power electronic inverter based generation (IBG) connected to the electric power distribution system (eg. roof top solar PV systems). This tool can be used to simulate the dynamics of a single DER connected to a stiff voltage source as shown in the following schematic:

![schematic of PV-DER](PVDER_schematic.png)

## Basics
The dynamics of the DER are modelled using dynamic phasors. Detailed description of the concepts behind this tool can be found in the IEEE publication [Dynamic Modeling of Solar PV Systems for Distribution System Stability Analysis](https://www.researchgate.net/publication/333985171_Dynamic_Modeling_of_Solar_PV_Systems_for_Distribution_System_Stability_Analysis) and detailed list of equations can be found in the [Model specification document.](docs/PV_DER_model_specification_rev3.docx)

### Features
The following features are available currently:
1. Single phase, three phase balanced, and three phase unbalanced (phase voltages may be unbalanced) DER models.
2. Run simulation in stand alone mode with internal grid voltage source (stiff) model.
3. Run simulation in loop mode where grid voltage is supplied every time step by a third party program.
4. Customize all aspects of the model through a [JSON](config_der.json) file which provides access to parameters in all the model components. 
5. Visualize or retrieve simulation results for voltages, current, active, and reactive power.
5. Introduce solar insolation events (in all modes), grid voltage, and frequency change events (in stand alone mode).
6. Retrieve and modify model parameters from a third party program.
7. Following smart inverter features are available: Low/High voltage ride through (LVRT/HVRT), Low frequency ride through (LFRT), and Volt-VAR control logic.

## Links
* Source code repository: https://github.com/sibyjackgrove/SolarPV-DER-simulation-tool
* API Documentation: https://solarpv-der-simulation-utility.readthedocs.io/en/latest/
* Additional documentation: [Description of attributes and methods](docs/PVDER_flags_variables_methods.md)

## Installation

Dependencies:

- SciPy >= 1.2.1
- Numpy >= 1.16.2
- Matlplotlib >= 3.0.3

Install latest release:
```
pip install pvder
```

Install from source:
```
git clone https://github.com/tdcosim/SolarPV-DER-simulation-tool.git
cd SolarPV-DER-simulation-tool
pip install -e .
```

## Use cases
Following projects are using Solar PV-DER simulation tool:
1. [Argonne Transmission and Distribution systems Co-Simulation tool (TDcoSim)](https://github.com/tdcosim/TDcoSim)
2. [OpenAI Gym Distributed Energy Resource Environment  (Gym-DER)](https://github.com/sibyjackgrove/gym-SolarPVDER-environment)

## Using the tool
This tool can be imported as a normal python module:

```python
import pvder
```

### Using the stand alone single phase DER model with 10 kW power rating
The following steps are required. Additional documentation on attributes and methods are available [here](docs/PVDER_flags_variables_methods.md).
1. First import the following classes:
```
from pvder.DER_components_single_phase import SolarPV_DER_SinglePhase
from pvder.grid_components import Grid
from pvder.dynamic_simulation import DynamicSimulation
from pvder.simulation_events import SimulationEvents
from pvder.simulation_utilities import SimulationResults
```
1. Create a **_SimulationEvents_** object: This object is used to add or remove disturbance events occurs during the simulation.
```
events = SimulationEvents()
```
2. Create a **Grid** object: This object describes the steady state model for the grid voltage source. It needs to be supplied with an **_SimulationEvents_** object.
```
grid = Grid(events=events)
```
3. Create a **SolarPV_DER_SinglePhase** or **SolarPV_DER_ThreePhase** object: This object describes the dynamic DER model. It needs both an **_SimulationEvents_** object, and a path name for JSON file containing the DER configuration parameters.  It also needs a **Grid** object in stand alone mode). Additionaly either the power rating of the DER or the id for the parameter dictionary should be provided.
```
PV_DER = SolarPV_DER_SinglePhase(events=events,configFile=r'config_der.json',gridModel=grid,derId= '10',standAlone = True)
```
4. Create a **DynamicSimulation** object: This object runs the simulation and stores the solution. It takes **_SimulationEvents_**, **Grid** and, **SolarPV_DER_SinglePhase** objects as arguments.
```
sim = DynamicSimulation(grid_model=grid,PV_model=PV_DER,events = events)
```
5. Create a **SimulationResults** object: This object is used to visualize the simulation results.
```
results = SimulationResults(simulation = sim)
```
6. Add an event (for e.g. solar insolation change at 10.0 s):
```
events.add_solar_event(10,90)
```
7. Specify simulation flags (for e.g. set the DEBUG_SIMULATION and DEBUG_POWER flag to true to observe the power at each time step.):
```
sim.DEBUG_SIMULATION = False
sim.DEBUG_POWER = False
```
8. Specify simulation stop time (for e.g. 20.0 s):
```
sim.tStop = 20.0
```
9. Run the simulation:
```
sim.run_simulation()
```
10. Visualize the results (for e.g. the power output at PCC-LV side):
```
results.PER_UNIT = False
results.plot_DER_simulation(plot_type='active_power_Ppv_Pac_PCC')
```

### Examples
Try out Jupyter notebooks with usage examples in Google Colab:

Basic usage:
[![Basic usage](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/sibyjackgrove/SolarPV-DER-simulation-tool/blob/master/examples/PV-DER_usage_example.ipynb)

Running simulation in loop mode:
[![Updating model parameters](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/sibyjackgrove/SolarPV-DER-simulation-tool/blob/master/examples/PV-DER_usage_example_loop_mode.ipynb)

Updating model parameters:
[![Updating model parameters](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/sibyjackgrove/SolarPV-DER-simulation-tool/blob/master/examples/PV-DER_parameter_update_example.ipynb)

Voltage anomaly, ride through, and momentary cessation:
[![Voltage anomaly](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/sibyjackgrove/SolarPV-DER-simulation-tool/blob/master/examples/PV-DER_usage_example_LVRT_momentary_cessation_with_recovery.ipynb)

Frequency anomaly, ride through, and trip:
[![Frequency anomaly](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/sibyjackgrove/SolarPV-DER-simulation-tool/blob/master/examples/PV-DER_usage_example_LFRT_with_trip.ipynb)

## Module details
A schematic of the relationship between differen classes in the module is shown in the figure below:
![schematic of software architecture](docs/software_architecture.png)

## Issues
Please feel free to raise an issue for bugs or feature requests.

## Who is responsible?

**Core developer:**
- Siby Jose Plathottam splathottam@anl.gov

**Contributor:**

- Karthikeyan Balasubramaniam kbalasubramaniam@anl.gov

## Acknowledgement

This project was supported by Kemal Celik, [U.S. DOE Office of Electricity, Solar Energy Technology Office](https://www.energy.gov/eere/solar/solar-energy-technologies-office) through the [SuNLaMP](https://www.energy.gov/eere/solar/sunshot-national-laboratory-multiyear-partnership-sunlamp) program.

The authors would like to acknowledge [Shrirang Abhyankar](https://github.com/abhyshr) and Puspal Hazra for their contribution.

## Citation
If you use this code please cite it as:
```
@misc{pvder,
  title = {{SolarPV-DER-simulation-tool}: A simulation tool for or solar photovoltaic distributed energy resources},
  author = "{Siby Jose Plathottam,Karthikeyan Balasubramaniam}",
  howpublished = {\url{https://github.com/sibyjackgrove/SolarPV-DER-simulation-tool}},
  url = "https://github.com/sibyjackgrove/SolarPV-DER-simulation-tool",
  year = 2019,
  note = "[Online; accessed 19-March-2019]"
}
```
### Copyright and License
Copyright © 2019, UChicago Argonne, LLC

Photovoltaic Distributed Energy Resource (PV-DER) Simulation tool is distributed under the terms of [BSD-3 OSS License.](LICENSE)


CHANGELOG.md
# Changelog

## 0.6.0 - 2023-09-27

### Bug fix

-   Removed support for Python 2
-   Changes to creating DER model instance using "DERModel" class.
    - Removed the need to specify "modelType" when creating a model instance. The model type is specified in the config file now.
	- "derID" is now a required argument.
	- "powerRating" and "VrmsRating" can only be supplied through the config file now.
-   Minor refactoring and code cleanup.
-   Added Dockerfile for using high-performance Julia ODE solvers library - Acknowledgement to Abel Siqueira (Netherlands eScience Center)

## 0.5.3 - 2021-09-13

### Bug fix

-   Removed pathlib module from setup.py

## 0.5.2 - 2021-09-09

### Bug fix

-   Added Numba to setup.py

## 0.5.1 - 2021-07-28

### Bug fix

-   Fixed bug due to logs folder not being found.

## 0.5.0 - 2021-07-23

### Features

-   Added current reference ramp rate limiter for all models.

### Bug fix

-   Fixed bug in HVRT logging.
-   Fixed bug that multiplied *m_limit* with *1e1* in the current control loop integral reset logic.
-   Changed controller gains for *ConstantVdc* models to make their response slightly underdamped. Previously the response was overdamped.

## 0.4.0 - 2021-06-02

### Features

-   Added minimum ride-through time before momentary cessation.
-   Added additional logging information during voltage anomalies.

### Bug fix

-   Minor bug fixes and code clean up


pvder/templates.py
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 30 10:19:16 2020

@author: splathottam
"""

single_phase_models = ["SolarPVDERSinglePhase","SolarPVDERSinglePhaseConstantVdc"]
three_phase_models = ["SolarPVDERThreePhase","SolarPVDERThreePhaseNoVrmsFilter","SolarPVDERThreePhaseConstantVdc","SolarPVDERThreePhaseBalanced","SolarPVDERThreePhaseNumba"]

constant_Vdc_models = ["SolarPVDERSinglePhaseConstantVdc","SolarPVDERThreePhaseConstantVdc"]

model_types = ['SinglePhase','SinglePhaseConstantVdc',
			   'ThreePhaseUnbalanced','ThreePhaseUnbalancedConstantVdc','ThreePhaseBalanced','ThreePhaseUnbalancedNumba','ThreePhaseUnbalancedNoVrmsFilter']

DER_design_template = {"SolarPVDERSinglePhase":
					   {"parent_config":"",
						"basic_specs":{'phases':('a'),'n_phases':1,'n_ODE':11,"model_type":"SolarPVDERSinglePhase"},
						"basic_options":{'t_stable':0.5,'m_steady_state':0.96,"Sinsol":100.0,"current_gradient_limiter":False,"Vrms_measurement_type":"minimum"},
					   "module_parameters":{"Np":2,"Ns":1000,"Vdcmpp0":750.0,"Vdcmpp_min": 650.0,"Vdcmpp_max": 800.0},
					   "inverter_ratings":{"Srated":10e3,"Vdcrated":550.0,"Ioverload":1.3,"Vrmsrated":177.0,"Iramp_max_gradient_real":1.0,"Iramp_max_gradient_imag":1.0},
					   "circuit_parameters":{"Rf_actual":0.002,"Lf_actual":25.0e-6,"C_actual":30.0e-6,"R1_actual":0.0019,"X1_actual":0.0561},
					   "controller_gains":{"Kp_GCC":12000.0,"Ki_GCC":4000.0,"Kp_DC":-4.0,"Ki_DC":-20.0,"Kp_Q":0.4,"Ki_Q":20.0,"wp": 20e4},
					   "steadystate_values":{"maR0":0.7,"maI0":0.0,"iaR0":0.5,"iaI0":0.01},
					   "initial_states":{"iaR":0,"iaI":0.0,"xaR":0.0,"xaI":0.0,"uaR":0.0,"uaI":0.0,
										 "Vdc":550.0,
										 "xDC":0.0,"xQ":0.0,
										 "xPLL":0.0,"wte":6.28}
					   
					   },
					   
					   "SolarPVDERSinglePhaseConstantVdc":
					   {"parent_config":"",
						"basic_specs":{'phases':('a'),'n_phases':1,'n_ODE':10,"model_type":"SolarPVDERSinglePhaseConstantVdc"},
						"basic_options":{'t_stable':0.5,'m_steady_state':0.96,"Sinsol":100.0,'use_Pref':False,"current_gradient_limiter":False,"Vrms_measurement_type":"minimum"},
					   "module_parameters":{"Np":2,"Ns":1000,"Vdcmpp0":750.0,"Vdcmpp_min": 650.0,"Vdcmpp_max": 800.0},
					   "inverter_ratings":{"Srated":10e3,"Vdcrated":550.0,"Ioverload":1.3,"Vrmsrated":177.0,"Iramp_max_gradient_real":1.0,"Iramp_max_gradient_imag":1.0},
					   "circuit_parameters":{"Rf_actual":0.002,"Lf_actual":25.0e-6,"R1_actual":0.0019,"X1_actual":0.0561},
					   "controller_gains":{"Kp_GCC":12000.0,"Ki_GCC":4000.0,"Kp_P":5000.0,"Ki_P":500.0,"Kp_Q":0.4,"Ki_Q":20.0,"wp": 20e4},
					   "steadystate_values":{"maR0":0.7,"maI0":0.0,"iaR0":0.5,"iaI0":0.01},
					   "initial_states":{"iaR":0,"iaI":0.0,"xaR":0.0,"xaI":0.0,"uaR":0.0,"uaI":0.0,
										 "xP":0.0,"xQ":0.0,
										 "xPLL":0.0,"wte":6.28}
					   },
					   
					   "SolarPVDERThreePhase":
					   {"parent_config":"",
						"basic_specs":{'phases':('a','b','c'),'n_phases':3,'n_ODE':24,'unbalanced':True,"model_type":"SolarPVDERThreePhase"},
						"basic_options":{'t_stable':0.5,'m_steady_state':0.96,"Sinsol":100.0,"current_gradient_limiter":False,"Vrms_measurement_type":"filtered"},#average,minimum,filtered
						"module_parameters":{"Np":11,"Ns":735,"Vdcmpp0":550.0,"Vdcmpp_min": 525.0,"Vdcmpp_max": 650.0}, 
						"inverter_ratings":{"Srated":50e3,"Vdcrated":550.0,"Ioverload":1.3,"Vrmsrated":177.0,"Iramp_max_gradient_real":1.0,"Iramp_max_gradient_imag":1.0},
					    "circuit_parameters":{"Rf_actual":0.002,"Lf_actual":25.0e-6,"C_actual":300.0e-6,"R1_actual":0.0019,"X1_actual":0.0561},
					    "controller_gains":{"Kp_GCC":6000.0,"Ki_GCC":2000.0,"Kp_DC":-2.0,"Ki_DC":-10.0,"Kp_Q":0.2,"Ki_Q":10.0,"wp": 20e4,"Tfilter_Vrms":0.1},
					    "steadystate_values":{"maR0":0.89,"maI0":0.0,"iaR0":1.0,"iaI0":0.001},
					    "initial_states":{"iaR":0,"iaI":0.0,"xaR":0.0,"xaI":0.0,"uaR":0.0,"uaI":0.0,
										 "ibR":0,"ibI":0.0,"xbR":0.0,"xbI":0.0,"ubR":0.0,"ubI":0.0,
										 "icR":0,"icI":0.0,"xcR":0.0,"xcI":0.0,"ucR":0.0,"ucI":0.0,
										 "Vdc":550.0,"xDC":0.0,"xQ":0.0,"xPLL":0.0,"wte":6.28}
					   
					   },
                       
                       "SolarPVDERThreePhaseNoVrmsFilter":
					   {"parent_config":"",
                        "basic_specs":{'phases':('a','b','c'),'n_phases':3,'n_ODE':23,'unbalanced':True,"model_type":"SolarPVDERThreePhaseNoVrmsFilter"},
						"basic_options":{'t_stable':0.5,'m_steady_state':0.96,"Sinsol":100.0,"current_gradient_limiter":False,"Vrms_measurement_type":"minimum"},
						"module_parameters":{"Np":11,"Ns":735,"Vdcmpp0":550.0,"Vdcmpp_min": 525.0,"Vdcmpp_max": 650.0}, 
						"inverter_ratings":{"Srated":50e3,"Vdcrated":550.0,"Ioverload":1.3,"Vrmsrated":177.0,"Iramp_max_gradient_real":1.0,"Iramp_max_gradient_imag":1.0},
					    "circuit_parameters":{"Rf_actual":0.002,"Lf_actual":25.0e-6,"C_actual":300.0e-6,"R1_actual":0.0019,"X1_actual":0.0561},
					    "controller_gains":{"Kp_GCC":6000.0,"Ki_GCC":2000.0,"Kp_DC":-2.0,"Ki_DC":-10.0,"Kp_Q":0.2,"Ki_Q":10.0,"wp": 20e4},
					    "steadystate_values":{"maR0":0.89,"maI0":0.0,"iaR0":1.0,"iaI0":0.001},
					    "initial_states":{"iaR":0,"iaI":0.0,"xaR":0.0,"xaI":0.0,"uaR":0.0,"uaI":0.0,
										 "ibR":0,"ibI":0.0,"xbR":0.0,"xbI":0.0,"ubR":0.0,"ubI":0.0,
										 "icR":0,"icI":0.0,"xcR":0.0,"xcI":0.0,"ucR":0.0,"ucI":0.0,
										 "Vdc":550.0,"xDC":0.0,"xQ":0.0,"xPLL":0.0,"wte":6.28}
					   },
					   
					   "SolarPVDERThreePhaseNumba":
					   {"parent_config":"",
						"basic_specs":{'phases':('a','b','c'),'n_phases':3,'n_ODE':23,'unbalanced':True,"model_type":"SolarPVDERThreePhaseNumba"},
						"basic_options":{'t_stable':0.5,'m_steady_state':0.96,"Sinsol":100.0,"current_gradient_limiter":False,"Vrms_measurement_type":"minimum"},
						"module_parameters":{"Np":11,"Ns":735,"Vdcmpp0":550.0,"Vdcmpp_min": 525.0,"Vdcmpp_max": 650.0}, 
						"inverter_ratings":{"Srated":50e3,"Vdcrated":550.0,"Ioverload":1.3,"Vrmsrated":177.0,"Iramp_max_gradient_real":1.0,"Iramp_max_gradient_imag":1.0},
					   "circuit_parameters":{"Rf_actual":0.002,"Lf_actual":25.0e-6,"C_actual":300.0e-6,"R1_actual":0.0019,"X1_actual":0.0561},
					   "controller_gains":{"Kp_GCC":6000.0,"Ki_GCC":2000.0,"Kp_DC":-2.0,"Ki_DC":-10.0,"Kp_Q":0.2,"Ki_Q":10.0,"wp": 20e4},
					   "steadystate_values":{"maR0":0.89,"maI0":0.0,"iaR0":1.0,"iaI0":0.001},
					   "initial_states":{"iaR":0,"iaI":0.0,"xaR":0.0,"xaI":0.0,"uaR":0.0,"uaI":0.0,
										 "ibR":0,"ibI":0.0,"xbR":0.0,"xbI":0.0,"ubR":0.0,"ubI":0.0,
										 "icR":0,"icI":0.0,"xcR":0.0,"xcI":0.0,"ucR":0.0,"ucI":0.0,
										 "Vdc":550.0,"xDC":0.0,"xQ":0.0,"xPLL":0.0,"wte":6.28}
					   
					   },
					   
					   
					   "SolarPVDERThreePhaseBalanced":
					   {"parent_config":"",
						"basic_specs":{'phases':('a','b','c'),'n_phases':3,'n_ODE':11,'unbalanced':False,"model_type":"SolarPVDERThreePhaseBalanced"},
						"basic_options":{'t_stable':0.5,'m_steady_state':0.96,"Sinsol":100.0,"current_gradient_limiter":False,"Vrms_measurement_type":"minimum"},
					   "module_parameters":{"Np":11,"Ns":735,"Vdcmpp0":550.0,"Vdcmpp_min": 525.0,"Vdcmpp_max": 650.0}, 
					   "inverter_ratings":{"Srated":50e3,"Vdcrated":550.0,"Ioverload":1.3,"Vrmsrated":177.0,"Iramp_max_gradient_real":1.0,"Iramp_max_gradient_imag":1.0},
					   "circuit_parameters":{"Rf_actual":0.002,"Lf_actual":25.0e-6,"C_actual":300.0e-6,"R1_actual":0.0019,"X1_actual":0.0561},
					   "controller_gains":{"Kp_GCC":6000.0,"Ki_GCC":2000.0,"Kp_DC":-2.0,"Ki_DC":-10.0,"Kp_Q":0.2,"Ki_Q":10.0,"wp": 20e4},
					   "steadystate_values":{"maR0":0.89,"maI0":0.0,"iaR0":1.0,"iaI0":0.001},
					   "initial_states":{"iaR":0,"iaI":0.0,"xaR":0.0,"xaI":0.0,"uaR":0.0,"uaI":0.0,
										 "Vdc":550.0,
										 "xDC":0.0,"xQ":0.0,
										 "xPLL":0.0,"wte":6.28}
						},
											  
					   "SolarPVDERThreePhaseConstantVdc":
					   {"parent_config":"",
						"basic_specs":{'phases':('a','b','c'),'n_phases':3,'n_ODE':22,'unbalanced':True,"model_type":"SolarPVDERThreePhaseConstantVdc"},
						"basic_options":{'t_stable':0.5,'m_steady_state':0.96,"Sinsol":100.0,"current_gradient_limiter":False,"Vrms_measurement_type":"minimum"},
						"module_parameters":{"Np":11,"Ns":735,"Vdcmpp0":550.0,"Vdcmpp_min": 525.0,"Vdcmpp_max": 650.0}, 
						"inverter_ratings":{"Srated":50e3,"Vdcrated":550.0,"Ioverload":1.3,"Vrmsrated":177.0,"Iramp_max_gradient_real":1.0,"Iramp_max_gradient_imag":1.0},
					   "circuit_parameters":{"Rf_actual":0.002,"Lf_actual":25.0e-6,"R1_actual":0.0019,"X1_actual":0.0561},
					   "controller_gains":{"Kp_GCC":6000.0,"Ki_GCC":2000.0,"Kp_P":500.0,"Ki_P":50.0,"Kp_Q":0.2,"Ki_Q":10.0,"wp": 20e4},
					   "steadystate_values":{"maR0":0.89,"maI0":0.0,"iaR0":1.0,"iaI0":0.001},
					   "initial_states":{"iaR":0,"iaI":0.0,"xaR":0.0,"xaI":0.0,"uaR":0.0,"uaI":0.0,
										 "ibR":0,"ibI":0.0,"xbR":0.0,"xbI":0.0,"ubR":0.0,"ubI":0.0,
										 "icR":0,"icI":0.0,"xcR":0.0,"xcI":0.0,"ucR":0.0,"ucI":0.0,
										 "xP":0.0,"xQ":0.0,
										 "xPLL":0.0,"wte":6.28}
					   }
					   }
	

VRT_config_template = {'LVRT':{'config_id':'',
								  'config':{'1':{'V_threshold':0.88,
												 't_threshold':5.0,
                                                 't_min_ridethrough':2,
												 'mode':'mandatory_operation',
												 't_start':0.0,
												 'threshold_breach':False},
											'2':{'V_threshold':0.7,
												 't_threshold':3.5,
                                                 't_min_ridethrough':1,
												 'mode':'mandatory_operation', #'momentary_cessation'
												 't_start':0.0,
												 'threshold_breach':False},
											'3':{'V_threshold':0.5,
												 't_threshold':1.0,
                                                 't_min_ridethrough':0.5,
												 'mode':'momentary_cessation',
												 't_start':0.0,
												 'threshold_breach':False},
								  }
						},
						'HVRT':{'config_id':'',
								  'config':{'1':{'V_threshold':1.06,
												't_threshold':1.0,
                                                't_min_ridethrough':0.5,
												'mode':'mandatory_operation',
												't_start':0.0,
												'threshold_breach':False},
										   '2':{'V_threshold':1.12,
												't_threshold':0.5,
                                                't_min_ridethrough':0.5,
												'mode':'mandatory_operation',
												't_start':0.0,
												'threshold_breach':False},
							  }
						 },
						'VRT_delays':{'config_id':'',
									 'config':{'output_cessation_delay':0.01,
											  'output_restore_delay':1.75,
											  'restore_Vdc':False}
					   }}

FRT_config_template =  {'LFRT':{'parent_config':'',
						   'config':{'1':{'F_LF':57.0,
										  't_LF_limit':1/60,
										  't_LFstart':0.0},
									 '2':{'F_LF':58.8,
										  't_LF_limit':299.0,
										  't_LFstart':0.0}
									}},
						'HFRT':{'parent_config':'',
						   'config':{'1':{'F_HF':61.2,
										  't_HF_limit':299.0,
										  't_HFstart':0.0},
									 '2':{'F_HF':62.0,
										  't_HF_limit':1/60,
										  't_HFstart':0.0}
									}},
						'FRT_delays':{'parent_config':'',
								 'config':{'FRT_INSTANTANEOUS_TRIP':False}}
						}						   
					  
#Voltage and frequency ride through settings from IEEE 1557-2018 Category III (Table 16, page 48) 
#V1 to V2 - zone 2,V1 < - zone 1  SolarPVDER_ThreePhaseConstantVdc
	
RT_config_template_old = {'LVRT':{'0':['V_threshold','t_threshold','mode','t_start','threshold_breach']},
					  'HVRT':{'0':['V_threshold','t_threshold','mode','t_start','threshold_breach']},
					  'OUTPUT_CESSATION_DELAY':'','OUTPUT_RESTORE_DELAY':'','RESTORE_Vdc':'',
					  'F_LF1':'','F_LF2':'','t_LF1_limit':'','t_LF2_limit':'',
					  'F_HF1':'','F_HF2':'',
					  't_HF1_limit':'','t_HF2_limit':'',
					  'FRT_INSTANTANEOUS_TRIP':''
					 }
